includes/class-magic-link.php

```diff
@@ -0,0 +1,801 @@
+
+		// phpcs:ignore WordPressVIPMinimum.Functions.RestrictedFunctions.wp_mail_wp_mail
+		$sent = \wp_mail(
+			$email['to'],
+			\wp_specialchars_decode( sprintf( $email['subject'], $blogname ) ),
+			$email['message'],
+			$email['headers']
+		);
+
+		return $sent;
+	}
+
+	/**
+	 * Verify and returns the valid token given a user, token and client.
+	 * 
+	 * This method cleans up expired tokens and returns the token data for
+	 * immediate use.
+	 *
+	 * @param int    $user_id User ID.
+	 * @param string $client  Client hash.
+	 * @param string $token   Token to verify.
+	 *
+	 * @return array|\WP_Error {
+	 *   Token data.
+	 *
+	 *   @type string $token  The token.
+	 *   @type string $client Client hash.
+	 *   @type string $time   Token creation time.
+	 * }
+	 */
+	public static function validate_token( $user_id, $client, $token ) {
+		$errors = new \WP_Error();
+		$user   = \get_user_by( 'id', $user_id );
+
+		if ( ! $user ) {
+			$errors->add( 'invalid_user', __( 'User not found.', 'newspack' ) );
+		} elseif ( ! self::can_magic_link( $user->ID ) ) {
+			$errors->add( 'invalid_user_type', __( 'Not allowed for this user', 'newspack' ) );
+		} else {
+			$tokens = \get_user_meta( $user->ID, self::TOKENS_META, true );
+			if ( empty( $tokens ) || empty( $token ) ) {
+				$errors->add( 'invalid_token', __( 'Invalid token.', 'newspack' ) );
+			}
+		}
+
+		$valid_token = false;
+
+		if ( ! $errors->has_errors() ) {
+			$expire = time() - self::get_token_expiration_period();
+
+			foreach ( $tokens as $index => $token_data ) {
+				if ( $token_data['time'] < $expire ) {
+					unset( $tokens[ $index ] );
+
+				} elseif ( $token_data['token'] === $token ) {
+					$valid_token = $token_data;
+
+					/** If token data has a client hash, it must be equal to the user's. */
+					if ( ! empty( $token_data['client'] ) && $token_data['client'] !== $client ) {
+						$errors->add( 'invalid_client', __( 'Invalid client.', 'newspack' ) );
+					}
+
+					unset( $tokens[ $index ] );
+					break;
+				}
+			}
+
+			if ( empty( $valid_token ) ) {
+				$errors->add( 'expired_token', __( 'Token has expired.', 'newspack' ) );
+			}
+			self::clear_client_secret_cookie();
+
+			$tokens = array_values( $tokens );
+			\update_user_meta( $user->ID, self::TOKENS_META, $tokens );
+		}
+
+		return $errors->has_errors() ? $errors : $valid_token;
+	}
+
+	/**
+	 * Handle a reader authentication attempt using magic link token.
+	 *
+	 * @param int    $user_id User ID.
+	 * @param string $token   Token to authenticate.
+	 *
+	 * @return bool|\WP_Error Whether the user was authenticated or WP_Error.
+	 */
+	private static function authenticate( $user_id, $token ) {
+		if ( \is_user_logged_in() ) {
+			return false;
+		}
+
+		$user = \get_user_by( 'id', $user_id );
+
+		if ( ! $user ) {
+			return new \WP_Error( 'invalid_user', __( 'User not found.', 'newspack' ) );
+		}
+
+		$client     = self::get_client_hash( $user );
```

Comment:
Suggestion: here and and in a few other instances, it'd be more readable if the result of `get_client_hash` was named `$client_hash`. If it's just `$client`, it lefts me wondering if it's maybe a PHP class or an object with data.

Reply:
Good call! Done in c61b4b8dcc192c7ab218924c5332ce4fbb49907c